pkg/agent/sysadvisor/plugin/qosaware/resource/memory/plugin/memory_provisioner.go

```diff
@@ -0,0 +1,126 @@
+
+	"github.com/kubewharf/katalyst-core/pkg/agent/qrm-plugins/memory/dynamicpolicy/memoryadvisor"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/metacache"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/plugin/qosaware/resource/memory/plugin/provisioner"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/types"
+	"github.com/kubewharf/katalyst-core/pkg/config"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+	"github.com/kubewharf/katalyst-core/pkg/util/machine"
+)
+
+func init() {
+	provisioner.RegisterInitializer(types.MemoryProvisionPolicyCanonical, provisioner.NewPolicyCanonical)
+}
+
+const (
+	MemoryProvisioner = "memory-provisioner"
+)
+
+type memoryProvisioner struct {
+	mutex            sync.RWMutex
+	metaReader       metacache.MetaReader
+	metaServer       *metaserver.MetaServer
+	emitter          metrics.MetricEmitter
+	memoryProvisions machine.MemoryDetails
+	conf             *config.Configuration
+	extraConfig      interface{}
+	// policy
+	policy provisioner.ProvisionPolicy
+}
+
+func NewMemoryProvisioner(conf *config.Configuration, extraConfig interface{}, metaReader metacache.MetaReader,
+	metaServer *metaserver.MetaServer, emitter metrics.MetricEmitter) MemoryAdvisorPlugin {
+	mp := &memoryProvisioner{
+		conf:             conf,
+		extraConfig:      extraConfig,
+		metaReader:       metaReader,
+		metaServer:       metaServer,
+		emitter:          emitter,
+		memoryProvisions: make(machine.MemoryDetails),
+	}
+	klog.Infof("[memory-provisioner] new memory provisioner")
+	if err := mp.initializeMemoryProvisioner(); err != nil {
+		general.Errorf("[memory-provisioner] initialize memory provisioner failed: %v", err)
+		return nil
+	}
+
+	return mp
+}
+
+// initializeMemoryProvisioner initialize memory provisioner
+func (m *memoryProvisioner) initializeMemoryProvisioner() error {
+	// get policy
+	policyName := m.conf.MemoryProvisionPolicy
+	initializers := provisioner.GetRegisteredInitializers()
+
+	initFunc, ok := initializers[policyName]
+	klog.Infof("[memory-provisioner] policyName: %s", policyName)
+	if !ok {
+		return fmt.Errorf("policy %s not found", policyName)
+	}
+	m.policy = initFunc(m.conf, nil, m.metaReader, m.metaServer, m.emitter)
+
+	return nil
+}
+
+func (m *memoryProvisioner) Reconcile(status *types.MemoryPressureStatus) (err error) {
+	reservedForAllocate := m.conf.GetDynamicConfiguration().
+		ReservedResourceForAllocate[v1.ResourceMemory]
+	m.policy.SetEssentials(
+		types.ResourceEssentials{
+			EnableReclaim:       m.conf.GetDynamicConfiguration().EnableReclaim,
+			ResourceUpperBound:  float64(m.metaServer.MemoryCapacity),
+			ReservedForAllocate: reservedForAllocate.AsApproximateFloat64(),
+		})
+	err = m.policy.Update()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *memoryProvisioner) GetAdvices() types.InternalMemoryCalculationResult {
```

Comment:
how will this `GetAdvices` be called?

Reply:
`memoryResourceAdvisor update` will call `sendAdvices()`, and `sendAdvices()` will call `GetAdvices()`.

https://github.com/kubewharf/katalyst-core/blob/4a3876efc7edd3f446c49912e559a8bcd0bab1f7/pkg/agent/sysadvisor/plugin/qosaware/resource/memory/advisor.go#L159-L165